pallets/confidential-asset/src/lib.rs

```diff
@@ -370,20 +404,247 @@ decl_module! {
+                account_id,
+                EncryptedBalanceWrapper::from(EncryptedAmount::default().encode()),
+            );
+        }
+
+        if <FailedOutgoingBalance>::contains_key(owner_did, account_id.clone()) {
+            <FailedOutgoingBalance>::insert(
+                owner_did,
+                account_id,
+                EncryptedBalanceWrapper::from(EncryptedAmount::default().encode()),
+            );
+        }
+    }
+
+    /// Add the `amount` to the mercat account balance, and update the `IncomingBalance` accumulator.
+    pub fn mercat_account_deposit_amount(
         owner_did: &IdentityId,
         account_id: &MercatAccountId,
-        new_encrypted_balance: EncryptedAmount,
-    ) {
+        amount: EncryptedAmount,
+    ) -> DispatchResult {
+        let current_balance =
+            Self::mercat_account_balance(owner_did, account_id.clone()).to_mercat::<T>()?;
+        <MercatAccountBalance>::insert(
+            owner_did,
+            account_id,
+            EncryptedBalanceWrapper::from((current_balance + amount).encode()),
+        );
+
+        // Update the incoming balance accumulator.
+        Ok(Self::add_incoming_balance(owner_did, account_id, amount)?)
+    }
+
+    /// Subtract the `amount` from the mercat account balance, and update the `PendingOutgoingBalance` accumulator.
+    pub fn mercat_account_withdraw_amount(
+        owner_did: &IdentityId,
+        account_id: &MercatAccountId,
+        amount: EncryptedAmount,
+    ) -> DispatchResult {
+        let current_balance =
+            Self::mercat_account_balance(owner_did, account_id.clone()).to_mercat::<T>()?;
         <MercatAccountBalance>::insert(
             owner_did,
             account_id,
-            EncryptedBalanceWrapper::from(new_encrypted_balance.encode()),
+            EncryptedBalanceWrapper::from((current_balance - amount).encode()),
+        );
+
+        // Update the pending balance accumulator.
+        let pending_balance =
+            Self::pending_outgoing_balance(owner_did, account_id.clone()).to_mercat::<T>()?;
+
+        <PendingOutgoingBalance>::insert(
+            owner_did,
+            account_id,
+            EncryptedBalanceWrapper::from((pending_balance - amount).encode()),
+        );
+        Ok(())
+    }
+
+    /// Calculate account owner's (transaction sender's) pending balance, given the previous
+    /// pending outgoing, incoming, and failed outgoing transaction amounts.
+    pub fn get_pending_balance(
+        owner_did: &IdentityId,
+        account_id: &MercatAccountId,
+    ) -> Result<EncryptedAmount, DispatchError> {
+        let mut current_balance =
+            Self::mercat_account_balance(owner_did, account_id.clone()).to_mercat::<T>()?;
+
+        if <PendingOutgoingBalance>::contains_key(owner_did, account_id.clone()) {
+            current_balance -=
+                Self::pending_outgoing_balance(owner_did, account_id.clone()).to_mercat::<T>()?;
+        }
```

Comment:
```suggestion
        if let Some(bal) = Self::pending_outgoing_balance(owner_did, account_id.clone()) {
            current_balance -= bal.to_mercat::<T>()?;
        }
```
and so on... (you'll need `Option<_>` in the storage items too)

Reply:
No difference in terms of perf, but it reduces duplication. Instead of checking for containment and then fetching, you just get `-> Option<Value>`. You have a logical mapping to `None` in either case; this change just manifests it in the typesystem, so its also more statically checked.

Additionally, you could chain these `Option`s together as an iterator, try_fold them with `.to_mercat::<T>()` and finally subtract once (makes the code somewhat cleaner).